designs/Bignums-design.md

```diff
@@ -0,0 +1,104 @@
+# Compiler A: Bignums - design
+
+## AST, IR, and Built-in Libraries
+- We will change `print` in the `built-in library` to print bignums from the heap.
+- We plan to modify the `built-in library` to have Javascript functions support all the unary and binary operations of bignums. This way we can first support the behavior of Python, and then improve the efficiency by writing the functions in WASM.
+
+## New Functions, Datatypes, and/or Files
+- We will change `codeGenValue` to allocate bignums in the memory. This will store the bignum in memory and leave the address on the stack.
```

Comment:
How will this interact with e.g. alloc/load/store so the memory management group knows what's going on with bignums? Remember that we don't have access to `$heap` anymore from generated code, only `memory.wat` has access to that.

Reply:
Does "interact with e.g. alloc/load/store" mean we have to know the size of the bignum we want to alloc before run-time? Shouldn't there be some way to reallocate a larger space in run-time like vectors in C++?

Reply:
We could make function calls to alloc/load/store in `memory.wat` to make runtime allocation.

Reply:
We could make function calls to alloc/load/store in `memory.wat` to make runtime allocation.